llpc/lower/llpcSpirvLowerTerminator.cpp

```diff
@@ -0,0 +1,176 @@
+// Initializes static members.
+char SpirvLowerTerminator::ID = 0;
+
+// =====================================================================================================================
+// Pass creator, creates the pass of SPIR-V lowering terminator operations
+ModulePass *createSpirvLowerTerminator() {
+  return new SpirvLowerTerminator();
+}
+
+// =====================================================================================================================
+SpirvLowerTerminator::SpirvLowerTerminator() : SpirvLower(ID) {
+}
+
+// =====================================================================================================================
+// Executes this SPIR-V lowering pass on the specified LLVM module.
+//
+// @param [in/out] module : LLVM module to be run on
+bool SpirvLowerTerminator::runOnModule(Module &module) {
+  LLVM_DEBUG(dbgs() << "Run the pass Spirv-Lower-Terminator\n");
+
+  SpirvLower::init(&module);
+
+  // Kills are only valid in fragment shader model.
+  if (m_shaderStage != ShaderStageFragment)
+    return false;
+
+  // Invoke handling of "kill" instructions.
+  visit(m_module);
+
+  // Remove any dead instructions.
+  bool changed = !m_removalStack.empty();
+  while (!m_removalStack.empty()) {
+    auto deadInst = m_removalStack.pop_back_val();
+    LLVM_DEBUG(dbgs() << "remove: " << *deadInst << "\n");
+    deadInst->dropAllReferences();
+    deadInst->eraseFromParent();
+  }
+  m_instsForRemoval.clear();
+
+  return changed;
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// Look for kills followed by instructions other than a return.
+// If found, mark dead instructions for removal and add a return immediately following the kill.
+//
+// @param callInst : "Call" instruction
+void SpirvLowerTerminator::visitCallInst(CallInst &callInst) {
+  auto callee = callInst.getCalledFunction();
+  if (!callee)
+    return;
+
+  auto mangledName = callee->getName();
+  if (mangledName != "lgc.create.kill")
+    return;
+
+  // Already marked for removal?
+  if (m_instsForRemoval.count(&callInst))
+    return;
+
+  BasicBlock *parentBlock = callInst.getParent();
+  auto instIter = std::next(callInst.getIterator());
+  assert(instIter != parentBlock->end() && "Should not be at end of block; illegal IR?");
+
+  // Already has a return?
+  if (isa<ReturnInst>(*instIter))
+    return;
+
+  // If terminator branches to block containing a PHI node which references
+  // this block then instructions following kill may be part of a return
+  // value sequence. In which case we cannot safely modify the instructions.
+  auto blockTerm = parentBlock->getTerminator();
+  if (!blockTerm)
+    return;
+  if (BranchInst *branch = dyn_cast<BranchInst>(blockTerm)) {
+    assert(branch->isUnconditional());
+    BasicBlock *targetBlock = branch->getSuccessor(0);
```

Comment:
Why we are limiting this only to conditional branches and to successor(0)/IfTrue ?
Shouldn't we check both successors for conditional branch ? And if we have unconditional branch then successor based on the condition ?

I'm asking because I found by accident test which is crashing on the assertion L151.
angle_deqp_khr_gles31_tests --use-angle=vulkan --gtest_filter=KHR-GLES31.core.shader_image_load_store.basic-allFormats-store-fs
@perlfu @nhaehnle